kythe/java/com/google/devtools/kythe/platform/java/filemanager/CompilationUnitPathFileManager.java

```diff
@@ -273,4 +349,42 @@ private void setSystemOption(String value) throws IOException {
       throw new IllegalArgumentException(value);
     }
   }
+
+  private FileObject readAhead(FileObject fo) {
+    Path path = asPath(fo);
+    if (path.getFileSystem().equals(fileSystem)) {
+      readAheadProvider.readAhead(path.toString(), path.toUri().getHost());
+    }
+    return fo;
+  }
+
+  private JavaFileObject readAhead(JavaFileObject fo) {
+    readAhead((FileObject) fo);
+    return fo;
+  }
+
+  private static class ReadAheadDataProvider implements FileDataProvider {
+    final Map<String, ListenableFuture<byte[]>> cache = new HashMap<>();
+    final FileDataProvider provider;
+
+    ReadAheadDataProvider(FileDataProvider provider) {
+      this.provider = provider;
+    }
+
+    void readAhead(String path, String digest) {
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(digest), "digest is empty");
+      cache.computeIfAbsent(digest, k -> provider.startLookup(path, digest));
+    }
+
+    @Override
+    public ListenableFuture<byte[]> startLookup(String path, String digest) {
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(digest), "digest is empty");
+      return cache.containsKey(digest) ? cache.remove(digest) : provider.startLookup(path, digest);
+    }
```

Comment:
So I find the near-duplication of readHead/startLookup a little odd. I get that you're trying to eagerly empty the cache. 

Is it a safe assumption that most files are read no more than once?

As an alternative to the remove on the call to startLookup, I think I'd rather see a callback added to the future, and have it removed from the cache on future completion.

That in itself points out that this should probably have some synchronization included to manage the cache.

Reply:
It depends on what you mean by "files are read no more than once".  The current indexer implementation reads a compilation unit's "source_files" at least once.

The callback approach seems to add unnecessary complexity to me; can you expand on the expected benefit?  There really isn't much code that can be shared between these two tiny functions without triggering an "error prone" failure for ignoring a returned future.

Reply:
Adding the Callback is a more idiomatic use of Futures. As for the ErrorProne warning, I'd actually prefer the @SuppressWarning to the 2nd method.

Regardless, of that, though, you need to handle concurrency here.

Reply:
Sure, I wasn't questioning the concurrency (although there isn't actually any concurrency involved outside of the Future, it's better to be safe).